install-deps.sh

```diff
@@ -237,6 +245,14 @@ else
                 elif test $(lsb_release -si) = VirtuozzoLinux -a $MAJOR_VERSION = 7 ; then
                     $SUDO yum-config-manager --enable cr
                 fi
+                $SUDO $yumdnf install -y py-bcrypt npm
+                if [ `npm -v | sed 's/\..*$//g'` -lt 5 ]; then
+                  $SUDO npm install -g yarn
+                  $SUDO yarn global add npm@5.0.3
+                  $SUDO npm uninstall -g yarn
+                  $SUDO mv /usr/bin/npm /usr/bin/npm.old
```

Comment:
Installing dependencies required for building certain packages on an OS is one thing, but I wonder if this is going a bit too far. A script called `install-deps` should not mess with the runtime environment and install packages outside of the control of the package management system.

Reply:
I agree with that, but in this case the alternative would be to rely in some third party URL to install a the version we need from npm. This way we are solely relying in npm to update its own version.
Do you see yet another alternative?

Reply:
I agree with @LenzGr. The alternative is alerting the user and have them taking the appropriate measures. Maybe create an argument like apt's "--yes", which will allow the script (if run from an automated environment, for instance) to do as it pleases.

Reply:
@LenzGr @jecluis although I agree with your concerns, a similar case already happens in the `install-deps.sh` script, see https://github.com/openattic/ceph/blob/wip-mgr-dashboard_v2/install-deps.sh#L148

I'm also not sure if adding a CLI option to `install-deps.sh` would not raise concerns to CI maintainers. Maybe we could call someone from upstream that is more or less responsible for this script, @jecluis who might that be?

Reply:
@rjfd a `pip install pecan` seems like something to be expected by an `install-deps.sh` script, as it is installing a dependency -- just through pip, instead of something else. The problem I personally have with this portion of the patch is not the `npm install`, but with moving the binaries (as that is changing the environment in a way that is not tracked by whatever package manager is on the box).

As for whom to contact, I'd say maybe Dan Mick or David Galloway.

Reply:
@jecluis thanks for clarifying the concern you have. I think if we just replace the `/usr/bin/npm` symlink instead of leaving a backup, then when the user uninstalls npm using the system package manager, everything will be cleaned. Don't you think?

Reply:
@rjfd no, I don't think we should be playing with the user's binaries in `/usr/bin/` at all, unless those binaries are Ceph's (and we have packages and `make install` for that). `/usr/bin/npm` should be created, managed, whatever, by whatever package provides `npm`, or by whatever the distro provides to handle alternatives (ubuntu does that via `update-alternatives` iirc, but I have no idea if `npm` is one of them, but given you're installing `npm` via `yarn`, I'm inclined to assume that would not be the case).

The matter, as I see it, is that I, as a developer of Ceph, do not want you to be messing around with my binaries in `/usr/bin`, unless I explicitly tell you to. Installing packages, sure, do that -- but that does not entail changing binaries themselves, as that's handled by my distro's package manager.

If you are dead set on having `npm` somewhere, and you want to control it, then I suggest adding that to the user's `$HOME/bin`, and letting the user know what you've done. Alternatively, spew out something like "please add $this directory to your PATH so you can have 'npm'", and then create the symlink or whatever on $this directory. Or just tell the user they need to do something to have a working npm. Whatever you decide to do, keep your hands off my `/usr/bin` :)

Reply:
@jecluis got it. Will try another way.